### DISCUSSION TITLE
Installation geht nicht

### DISCUSSION THREAD

**Discussion by Kasper027** [openv/openv#535] [General]
Hallo zusammen, 
ich versuche mich gerade an der Anleitung zur Installation des openv auf einem Raspberry. Leider habe ich es nicht richtig verstanden. 

Zum Vorgehen:
Die ersten Schritte sind soweit klar. Ich habe den Ordner angelegt, geöffnet und dort das `git clone https://github.com/openv/vcontrold.git
` ausgeführt. 
Das hat alles geklappt. Danach laut Anleitung `sudo cmake ./vcontrold/ -DVSIM=ON -DMANPAGES=OFF` ausgeführt mit folgendem 

> -- The C compiler identification is GNU 10.2.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Found LibXml2: /usr/lib/arm-linux-gnueabihf/libxml2.so (found version "2.9.10")
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Configuring done
-- Generating done
-- Build files have been written to: /home/pi/openv

Das scheint ja erstmal alle geklappt zu haben. Dich wie geht es nun weiter?
Ich habe es mit `sudo make install` versucht. Doch das bricht er ab mit fehlern:

> Scanning dependencies of target UpdateVersion
[  4%] Updating version header.
-- Updating version information to 0.98.10-23-g600f59e ...
[  4%] Built target UpdateVersion
Scanning dependencies of target vclient
[  9%] Building C object CMakeFiles/vclient.dir/src/common.c.o
[ 14%] Building C object CMakeFiles/vclient.dir/src/socket.c.o
[ 19%] Building C object CMakeFiles/vclient.dir/src/io.c.o
[ 23%] Building C object CMakeFiles/vclient.dir/src/client.c.o
[ 28%] Building C object CMakeFiles/vclient.dir/src/vclient.c.o
[ 33%] Linking C executable vclient
[ 33%] Built target vclient
Scanning dependencies of target vsim
[ 38%] Building C object CMakeFiles/vsim.dir/src/socket.c.o
[ 42%] Building C object CMakeFiles/vsim.dir/src/vsim.c.o
[ 47%] Linking C executable vsim
[ 47%] Built target vsim
Scanning dependencies of target vcontrold
[ 52%] Building C object CMakeFiles/vcontrold.dir/src/io.c.o
[ 57%] Building C object CMakeFiles/vcontrold.dir/src/common.c.o
[ 61%] Building C object CMakeFiles/vcontrold.dir/src/xmlconfig.c.o
[ 66%] Building C object CMakeFiles/vcontrold.dir/src/parser.c.o
[ 71%] Building C object CMakeFiles/vcontrold.dir/src/socket.c.o
[ 76%] Building C object CMakeFiles/vcontrold.dir/src/semaphore.c.o
[ 80%] Building C object CMakeFiles/vcontrold.dir/src/framer.c.o
[ 85%] Building C object CMakeFiles/vcontrold.dir/src/unit.c.o
[ 90%] Building C object CMakeFiles/vcontrold.dir/src/arithmetic.c.o
[ 95%] Building C object CMakeFiles/vcontrold.dir/src/vcontrold.c.o
[100%] Linking C executable vcontrold
/usr/bin/ld: cannot open output file vcontrold: Ist ein Verzeichnis
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/vcontrold.dir/build.make:239: vcontrold] Fehler 1
make[1]: *** [CMakeFiles/Makefile2:155: CMakeFiles/vcontrold.dir/all] Fehler 2
make: *** [Makefile:149: all] Fehler 2

Kann mir jemand helfen? 
Viele Grüße
Kasper027

**Comment by speters** [comment]
> Das hat alles geklappt. Danach laut Anleitung sudo cmake ./vcontrold/ -DVSIM=ON -DMANPAGES=OFF ausgeführt ...

In welcher Anleitung steht das denn so?
So wie du das machst, versucht der Linker das `vcontrold` Programm in einem Verzeichnis zu erzeugen, in welchem es bereits ein Unterverzeichnis mit gleichem Namen gibt (_/usr/bin/ld: cannot open output file vcontrold: Ist ein Verzeichnis_). Das geht natürlich so nicht.

Fang nochmal frisch an und mach es so, wie in [INSTALL.md](https://github.com/openv/vcontrold/blob/master/doc/INSTALL.md) beschrieben (dort werden die Sachen in einem `build` genannten Unterverzeichnis des Quellcodeordners gebaut).

**Reply by Kasper027** [reply]
Hey @speters 

Ich bin hiernach vorgegangen. Nur leider ist der teil 3 ein wenig unübersichtlich. 
[https://github.com/openv/openv/wiki/vcontrold-mit-Raspberry-Pi](https://github.com/openv/openv/wiki/vcontrold-mit-Raspberry-Pi)

Dann nur noch mal für mich zum besseren Verständniss:
1. ich lade den Quellcode herunter
2. darin erstelle ich den Ordner builds und öffne diesen
3. dann nur den Befehl `cmake -DVSIM=ON -DMANPAGES=OFF` ausführen
4. `sudo make install` ausführen

So richtig?

**Reply by speters** [reply]
Zu 3.:
`cmake -DVSIM=ON -DMANPAGES=OFF ..` im ./build Verzeichnis.
Beachte die "..". Das weist CMake an, im Elternverzeichnis nach dem CMakeLists.txt zu suchen.
Dieser Schritt erstellt und konfiguriert die Makefiles.

Dann ein `make`, um die Programme zu kompilieren. Diesen Schritt gerne ohne _sudo_, da hier noch keine root-Rechte benötigt werden und man aus Sicherheitsgründen die Benutzung auf das Mindeste einschränkt.

Zu 4.:
Du kannst bei sowas vorher mit `make -n install` einen Probelauf machen, um zu sehen, was da überhaupt installiert werden soll. Grundsätzlich könnten in einem Makefile nämlich auch Anweisungen für fiese Dinge drinstehen (stell Dir mal ein mit root-Rechten ausgeführtes `rm -rf /` als Installschritt vor).

**Reply by besimo1** [reply]
Nach EIngabe "cmake -DVSIM=ON -DMANPAGES=OFF .." im ./build Verzeichnis.
Kommt bei mir folgender Fehler:
CMake Error: The source directory "/home/pi/openv" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.

Was mache ich falsch ? (die vorherigen Schritte habe ich alle gemacht)

**Comment by Kasper027** [comment]
Hey. Vielen Dank erstmal. Die Installation hat schonmal fast komplett geklappt. Muss jetzt nur noch es so hin bekommen, dass es automatisch auch startet. Aber das wird dann auch noch. Ach Ja und dann noch alles konfigurieren :-)

**Comment by Kasper027** [comment]
Hey @speters,
hoffe, du kannst mir nochmal weiter helfen. Ich habe es nun soweit am laufen. Habe nun aber das Problem, dass der Service nicht automatisch startet. Vllt kannst du mir nochmal weiter helfen.

Also mit `vcontrol -n` läuft alles und ich kann mich mit einer Telnet verbinden.
Danach habe ich mit dem Befehl `sudo nano /etc/init.d/vcontrol` das Startupscript erstellt und diesen Code eingefügt

```
#! /bin/sh
  ### BEGIN INIT INFO
  # Provides:          vcontrold
  # Required-Start:    $remote_fs $syslog
  # Required-Stop:     $remote_fs $syslog
  # Default-Start:     2 3 4 5
  # Default-Stop:      0 1 6
  # Short-Description: Initscript to start vcontrold deamon
  # Description:       This file should be used to construct scripts to be
  #                    placed in /etc/init.d.
  ### END INIT INFO
  
  # Author:       Michael Pucher <tech@michaelpucher.net>
  # Homepage:     openv.wikispaces.org
  #
  # Please remove the "Author" lines above and replace them
  # with your own name if you copy and modify this script.

# Do NOT "set -e"

# PATH should only include /usr/* if it runs after the mountnfs.sh script
PATH=/sbin:/usr/sbin:/bin:/usr/bin
DESC="vcontrold deamon"
NAME=vcontrold
DAEMON=/usr/local/sbin/$NAME
#DAEMON_ARGS="--options args"
DAEMON_ARGS=""
PIDFILE=/var/run/$NAME.pid
SCRIPTNAME=/etc/init.d/$NAME

# Exit if the package is not installed
[ -x "$DAEMON" ] || exit 0

# Read configuration variable file if it is present
[ -r /etc/default/$NAME ] && . /etc/default/$NAME

# Load the VERBOSE setting and other rcS variables
. /lib/init/vars.sh

# Define LSB log_* functions.
# Depend on lsb-base (>= 3.2-14) to ensure that this file is present
# and status_of_proc is working.
. /lib/lsb/init-functions

#
# Function that starts the daemon/service
#
do_start()
{
    # Return
    #   0 if daemon has been started
    #   1 if daemon was already running
    #   2 if daemon could not be started
    start-stop-daemon --start --quiet --pidfile $PIDFILE --exec $DAEMON --test > /dev/null \
        || return 1
    start-stop-daemon --start --quiet --pidfile $PIDFILE --exec $DAEMON $DAEMON_ARGS \        || return 2
    # Add code here, if necessary, that waits for the process to be ready
    # to handle requests from services started subsequently which depend
    # on this one.  As a last resort, sleep for some time.
}

#
# Function that stops the daemon/service
#
do_stop()
{
    # Return
    #   0 if daemon has been stopped
    #   1 if daemon was already stopped
    #   2 if daemon could not be stopped
    #   other if a failure occurred
    start-stop-daemon --stop --quiet --retry=TERM/30/KILL/5 --pidfile $PIDFILE --name $NAME
    RETVAL="$?"
    [ "$RETVAL" = 2 ] && return 2
    # Wait for children to finish too if this is a daemon that forks
    # and if the daemon is only ever run from this initscript.
    # If the above conditions are not satisfied then add some other code
    # that waits for the process to drop all resources that could be
    # needed by services started subsequently.  A last resort is to
    # sleep for some time.
    start-stop-daemon --stop --quiet --oknodo --retry=0/30/KILL/5 --exec $DAEMON
    [ "$?" = 2 ] && return 2
    # Many daemons don't delete their pidfiles when they exit.
    rm -f $PIDFILE
    return "$RETVAL"
}

#
# Function that sends a SIGHUP to the daemon/service
#
do_reload() {
    #
    # If the daemon can reload its configuration without
    # restarting (for example, when it is sent a SIGHUP),
    # then implement that here.
    #
    start-stop-daemon --stop --signal 1 --quiet --pidfile $PIDFILE --name $NAME
    return 0
}

case "$1" in
  start)
    [ "$VERBOSE" != no ] && log_daemon_msg "Starting $DESC" "$NAME"
    do_start
    case "$?" in
        0|1) [ "$VERBOSE" != no ] && log_end_msg 0 ;;
        2) [ "$VERBOSE" != no ] && log_end_msg 1 ;;
    esac
    ;;
  stop)
    [ "$VERBOSE" != no ] && log_daemon_msg "Stopping $DESC" "$NAME"
    do_stop
    case "$?" in
        0|1) [ "$VERBOSE" != no ] && log_end_msg 0 ;;
        2) [ "$VERBOSE" != no ] && log_end_msg 1 ;;
    esac
    ;;
  status)
    status_of_proc "$DAEMON" "$NAME" && exit 0 || exit $?
    ;;
  #reload|force-reload)
    #
    # If do_reload() is not implementedthen leave this commented out
    # and leave 'force-reload' as an alias for 'restart'.
    #
    #log_daemon_msg "Reloading $DESC" "$NAME"
    #do_reload
    #log_end_msg $?
    #;;
  restart|force-reload)
    #
    # If the "reload" option is implemented then remove the
    # 'force-reload' alias
    #
    log_daemon_msg "Restarting $DESC" "$NAME"
    do_stop
    case "$?" in
      0|1)
        do_start
        case "$?" in
            0) log_end_msg 0 ;;
            1) log_end_msg 1 ;; # Old process is still running
            *) log_end_msg 1 ;; # Failed to start
        esac
        ;;
      *)
        # Failed to stop
        log_end_msg 1
        ;;
    esac
    ;;
  *)
    #echo "Usage: $SCRIPTNAME {start|stop|restart|reload|force-reload}" >&2
    echo "Usage: $SCRIPTNAME {start|stop|status|restart|force-reload}" >&2
    exit 3
    ;;
esac

:
```
 
Danach habe ich `sudo chmod u+x vcontrol` ausgeführt.
Wenn ich nun `sudo update-rc.d vcontrol start 99 2 3 4 5 . stop 99 0 1 6 .` ausführe, kommt folgendes:
> sudo update-rc.d vcontrol start 99 2 3 4 5 . stop 99 0 1 6 .
update-rc.d: warning: start and stop actions are no longer supported; falling back to defaults

Aber auch nach einem Neustart oder so kann ich mich nicht mit Telnet verbinden. Kannst du mir sagen, wie ich den Service starte?

**Comment by speters** [comment]
Bei einem Init-System (mittlerweile wird sehr oft SystemD verwendet) wird mit dem Befehl `sudo update-rc.d vcontrold defaults` das automatische Starten und Stoppen für die voreingestellten Runlevels eingetragen (im Prinzip durch Anlegen von Symlinks).

Jetzt nochmal mit `ps xawu | grep vcontrold` schauen, ob noch übergebliebene Prozesse von früheren vcontrold-Startversuchen laufen und diese gegebenenfalls mit `sudo killall -9 vcontrold` bzw. `sudo kill -9 PROZESSID` beenden.

Nun kannst du mit `sudo service vcontrold start` starten. Ausserdem wird _vcontrold_ bei einem Systemstart automatisch gestartet.

**Comment by ManuelK83** [comment]
Hallo ich klinke mich hier mal ein.

Ich habe auch das Problem, das der Service nicht läuft.  Bei mir ist vcontrold abe auf einem LXC container installiert.  Ich habe SystemD benutzt und folgendes in den Service geschrieben.
`[Unit]
Description=vcontrold daemon
After=syslog.target systemd-udev-settle.service

[Service]
Type=forking
ExecStartPre=/bin/ls /dev/bus/usb/ttyUSB0
ExecStart=/usr/local/sbin/vcontrold -x /etc/vcontrold/vcontrold.xml
ExecReload=/bin/kill -HUP $MAINPID
Restart=always
RestartSec=120
StandardOutput=null

[Install]
WantedBy=multi-user.target` 

Es funktioniert sogar einige Zeit.  Ich starte Proxmox und dann den LXC Container mit vcontrold.  Die Verbindung  funktioniert manchmal einige Minuten manchmal Stunden.    Dann klappt plötzlich nichts mehr.   Ich muss erst  komplett alles wieder neu starten.  
Ich dachte erst es liegt an der USB Verbindung.   Aber im LXC container aktuell  ist der ttyUSB0  immer noch aktiv. Vcontrold aber nicht mehr. 

Ich glaube eher der Service funktioniert nicht richtig.  Ich habe leider keine Ahnung wie das funktioniert. SystemD  auch nur kopiert.

Hoffe es kann mir einer helfen wie das richtig funktioniert.
